### PR TITLE
Add S7 Elephant Rev2 Support

### PR DESCRIPTION
**Pull request by MudkipMao**
Opening again because I really borked the last pull request.

S7 Elephant had a GB in late 2019 that is now shipping to customers. It features a new PCB with backlight compatibility. This change aims to bring QMK compatibility as well as allowing for VIA compatibility in the future.

## Description

I refactored the whole folder into a rev1 and rev2 scheme since the MCU pinouts are very different between the two PCB's. I tried to leave rev1 intact but do not have a PCB to test with. Rev2 is tested on my own PCB. Additionally, I have laid out the groundwork for VIA support which MechMerlin has finished.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## What needs work
* Finishing VIA support
* Adding more layouts, rev1 has many more layouts to make it easier for people to customize. Currently rev2 only has LAYOUT_all and LAYOUT_ansi

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/s7_elephant/rev2/keymaps/default/keymap.c [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+void led_set_user(uint8_t usb_led) {
+    // Turn LEDs On/Off for Caps Lock
+    if (host_keyboard_leds() & (1 << USB_LED_CAPS_LOCK)) {
+        rgblight_enable_noeeprom();
+        rgblight_sethsv_noeeprom(0, 0, 80);
+    } else {
+        rgblight_sethsv_noeeprom(0, 0, 80);
+        rgblight_disable_noeeprom();
+    }
+}
```
Comment (fauxpark) [inline keyboards/s7_elephant/rev2/keymaps/default/keymap.c:64]:
```suggestion
bool led_update_user(led_t led_state) {
    // Turn LEDs On/Off for Caps Lock
    if (led_state.caps_lock) {
        rgblight_enable_noeeprom();
        rgblight_sethsv_noeeprom(0, 0, 80);
    } else {
        rgblight_sethsv_noeeprom(0, 0, 80);
        rgblight_disable_noeeprom();
    }
    return false;
}
```

File: keyboards/s7_elephant/rev2/keymaps/via/rules.mk [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# Build options
+
+VIA_ENABLE = yes         # Enable usage of VIA configurator
+BOOTMAGIC_ENABLE = lite  # Enable bootmagic lite, holding a key while plugging in USB will jump to bootloader and reset EEPROM
```
Comment (fauxpark) [inline keyboards/s7_elephant/rev2/keymaps/via/rules.mk:4]:
```suggestion
VIA_ENABLE = yes
```

File: keyboards/s7_elephant/rev2/rules.mk [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no             # USB Nkey Rollover
+BACKLIGHT_ENABLE = yes       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes        # Enable keyboard RGB underglow
+MIDI_ENABLE = no             # MIDI support
+UNICODE_ENABLE = no          # Unicode
+BLUETOOTH_ENABLE = no        # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no            # Audio output on port C6
+FAUXCLICKY_ENABLE = no       # Use buzzer to emulate clicky switches
+HD44780_ENABLE = no          # Enable support for HD44780 based LCDs
```
Comment (fauxpark) [inline keyboards/s7_elephant/rev2/rules.mk:33]:
```suggestion
```

**Comment by sigprof** [comment]
Would adding `keyboards/s7_elephant/rules.mk` with
```make
DEFAULT_FOLDER = s7_elephant/rev1
```
and maybe `keyboards/s7_elephant/s7_elephant.h` with
```c
#pragma once

#include "quantum.h"

#if defined(KEYBOARD_s7_elephant_rev1)
#    include "rev1.h"
#elif defined(KEYBOARD_s7_elephant_rev2)
#    include "rev2.h"
#endif
```
be enough to merge this without going through the breaking changes process and the associated delay of several months? People are already coming to the QMK Discord after bricking their rev2 PCBs by flashing the rev1 firmware.

**Comment by Erovia** [comment]
I, personally, would like to see as much shared code between the 2 versions as possible.  
Afaik, the layout of the 2 versions are identical, so see no reason to not have a shared `keymaps` dir.

Kinda something what's going on with Planck and its revisions.

**Comment by MudkipMao** [comment]
@Erovia My initial commit was like that, but I was told to follow the revision system of the kbd67 which shares no code between versions. As long as we can agree on a standard, I don't mind doing either

**Comment by noroadsleft** [comment]
> Would adding `keyboards/s7_elephant/rules.mk` ...  be enough to merge this without going through the breaking changes process and the associated delay of several months?

My intention was actually to merge this today (note the target branch `kb_breaking/2020-06-13`), which I still may do depending on how quickly we get this sorted. I realized during the last breaking changes cycle that single-keyboard updates like this one wasted a lot of time in limbo during the three months between `future` (now `develop`) merges, and could benefit from a more rapid timeline.

Using `DEFAULT_FOLDER` isn't my preference, and perhaps I inadvertently let some of my bias against using it slip through when I gave @MudkipMao instructions regarding updating this codebase on the previous PR. Eventually, I'd like to do away with `DEFAULT_FOLDER` entirely, but the idea of using it completely slipped my mind.

All this is to say: "Yes, we could do that, and maybe we should."

**Comment by MudkipMao** [comment]
Ok @noroadsleft, I just read this. Should I put a DEFAULT_FOLDER in a rules.mk in the base folder?

**Comment by noroadsleft** [comment]
> Should I put a DEFAULT_FOLDER in a rules.mk in the base folder?

Yes. Just the one line and the one file will do.

**Comment by MudkipMao** [comment]
Ok I have done that

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**